packages/assets-controllers/src/TokenDetectionController.ts

```diff
@@ -318,10 +339,17 @@ export class TokenDetectionController extends StaticIntervalPollingController<To
       'TokenListController:stateChange',
       // TODO: Either fix this lint violation or explain why it's necessary to ignore.
       // eslint-disable-next-line @typescript-eslint/no-misused-promises
-      async ({ tokenList }) => {
-        const hasTokens = Object.keys(tokenList).length;
-
-        if (hasTokens) {
+      async ({ tokensChainsCache }) => {
+        // compare previous and incoming tokensChainsCache
+        const previousTokensChainsCache = removeTimestamps(
+          this.#tokensChainsCache,
+        );
+        const cleanTokensChainsCache = removeTimestamps(tokensChainsCache);
+        const isEqualValues = isEqual(
+          cleanTokensChainsCache,
+          previousTokensChainsCache,
+        );
+        if (!isEqualValues) {
```

Comment:
I wanted to compare not only the data but also the keys of the incoming tokensChainsCache.

Inside tokenListController, we would still make state [updates](https://github.com/MetaMask/core/blob/35b900bb779c27807e84254b36a153050a3ce426/packages/assets-controllers/src/TokenListController.ts#L337) (to update the timestamp) even when we are still fetching tokenList from cache, which will trigger the [tokenListController:stateChange](https://github.com/MetaMask/core/blob/35b900bb779c27807e84254b36a153050a3ce426/packages/assets-controllers/src/TokenDetectionController.ts#L317)

In the scenario when a user is switching back and forth between ethereum and linea for example; 
The detectTokens is triggered by the NetworkController:networkDidChange in [tokenDetectionController](https://github.com/MetaMask/core/blob/35b900bb779c27807e84254b36a153050a3ce426/packages/assets-controllers/src/TokenDetectionController.ts#L366).
and also triggered by the state updates in tokenListController even if we only have two networks and the tokenList is exactly the same (just the timestamp is being updated each time)

I wanted to avoid that one extra call ☝️  and instead make sure that only 1 call to detectTokens is made.

But at the same time, i'm aware that when a new entry is added to the list like polygon `0x89`, this will trigger the `detectTokens` function call; i thought that is fine because we will still be making less calls by ignoring the timestamp updates;

Reply:
Thank you @kanthesha for clarifying! Agree with that, updated here [2d2182d](https://github.com/MetaMask/core/pull/4821/commits/2d2182d1a7169a027db1b73779b76240ab44c5ac)